File: classes/game/item/GeneratorBasedItem.cpp
#include "GeneratorBasedItem.h"
#include "../Cell.h"
#include "../Level.h"
#include "../FallManager.h"
#include "../Mathf.h"
#include "../ItemGeneratorController.h"

void GeneratorBasedItem::StartGeneratingItems() {
	this->GenerateItems();
}

/*IEnumerator*/void GeneratorBasedItem::GenerateItems() {
	bool shouldPlayThrowAnimation = false;
//case 0:
	if(!this->_waiting) {
		this->_waiting = true;
		FallManager::IncreaseWaitingFallBlocker();
		shouldPlayThrowAnimation = false;
//		this->$current = null;
//		if(!this->$disposing) {
//			this->$PC = 1;
//		}
//		goto nextframe;
	}
	else
		return;
//	goto end;
//
//case 1:
	int availableCountOnBoard = this->CurrentLevel->CanGenerateCount(this->GenerateItemType);
	int availableCount = Mathf::Min(availableCountOnBoard, this->GenerateCount);
	if(availableCount > 0) {
		if(availableCountOnBoard <= 1) {
			ItemGeneratorController::getInstance()->SleepGenerators(this->GenerateItemType);
		}
		this->IntervalCounter++;
		if((this->IntervalCounter % this->GenerateInverval) == 0) {
			this->IntervalCounter = 0;
			this->_targetCellList.clear();
			for(int i = 0; i < availableCount; i++) {
				Cell* randomCell = this->GetRandomCell();
				if(randomCell == nullptr) {
					break;
				}
				this->_targetCellList.push_back(randomCell);
				this->CurrentLevel->IncreaseItemCount(this->GenerateItemType);
				randomCell->ReserveForItem(this->GenerateItemType);
				shouldPlayThrowAnimation = true;
			}
		}
//		break;
// [breakTo]

		if(shouldPlayThrowAnimation && (this->_targetCellList.size() > 0)) {
			this->PlayThrowAnimation(&this->_targetCellList);
		}
		this->_waiting = false;
		FallManager::DecreaseWaitingFallBlocker();
		//this->$current = null;
		//if(!this->$disposing) {
		//	this->$PC = 2;
		//}
		//goto nextframe;
	}
	else
		FallManager::DecreaseWaitingFallBlocker();
//	goto end;
//
//case 2:
//	this->$PC = -1;
//	goto end;
//
//default:
//	goto end;
//}
//end:
//return false;
//nextframe:
//return true;
}

void GeneratorBasedItem::OnThrowAnimationComplete(Cell* targetCell) {
	if (targetCell != nullptr) {
		targetCell->ApplyReservedItem();
	}
	this->CurrentLevel->DecreaseItemCount(this->GenerateItemType);
}

File: doc/porting/GeneratorBasedItem.cs
using Assets.Scripts.GamePlayScene.Mechanics;
using Assets.Scripts.GamePlayScene.Mechanics.Items;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics;

namespace GamePlayScene.Mechanics.Items
{
	public abstract class GeneratorBasedItem : Item
	{
		protected ItemType GenerateItemType;

		protected int GenerateCount = 1;

		protected int GenerateInverval = 1;

		protected int IntervalCounter;

		private readonly List<Cell> _targetCellList = new List<Cell>();

		private bool _waiting;

		public override ItemType GetItemType()
		{
			throw new NotImplementedException();
		}

		public override void ChangeSorting(Sorting s)
		{
			throw new NotImplementedException();
		}

		public override Sorting GetSorting()
		{
			throw new NotImplementedException();
		}

		public abstract void Sleep();

		public abstract Cell GetRandomCell();

		public abstract void PlayThrowAnimation(List<Cell> targetCells);

		protected void StartGeneratingItems()
		{
			base.StartCoroutine(this.GenerateItems());
		}

		[DebuggerHidden]
		private IEnumerator GenerateItems()
		{
			GeneratorBasedItem.<GenerateItems>c__Iterator0 <GenerateItems>c__Iterator = new GeneratorBasedItem.<GenerateItems>c__Iterator0();
			<GenerateItems>c__Iterator.$this = this;
			return <GenerateItems>c__Iterator;
		}

		protected virtual void OnThrowAnimationComplete(Cell targetCell)
		{
			if (targetCell != null)
			{
				targetCell.ApplyReservedItem();
			}
			this.CurrentLevel.DecreaseItemCount(this.GenerateItemType);
		}
	}
}
